cmd/vault-subpath-proxy/kv_update_transport.go

```diff
@@ -96,9 +119,124 @@ func (k *kvUpdateTransport) RoundTrip(r *http.Request) (*http.Response, error) {
 	} else {
 		go k.syncSecret(body.Data)
 	}
+	if response.StatusCode > 199 && response.StatusCode < 300 {
+		path := strings.TrimPrefix(r.URL.Path, "/v1/")
+		k.updateKeyCacheForSecret(path, body.Data)
+
+	}
 	return response, nil
 }
 
+func (k *kvUpdateTransport) updateKeyCacheForSecret(path string, item map[string]string) {
+	k.existingSecretKeysByNamespaceNameLock.Lock()
+	defer k.existingSecretKeysByNamespaceNameLock.Unlock()
+	fmt.Printf("Updating cache for secret at path '%s'\n", path)
+
+	// Clear old entries associated with us
+	for _, existingEntry := range k.existingSecretKeysByVaultSecretName[path] {
+		k.existingSecretKeysByNamespaceName[existingEntry.name].Delete(existingEntry.key)
+	}
+	delete(k.existingSecretKeysByVaultSecretName, path)
+
+	// Create new entries
+	name := types.NamespacedName{Namespace: item[vault.SecretSyncTargetNamepaceKey], Name: item[vault.SecretSyncTargetNameKey]}
+	if k.existingSecretKeysByNamespaceName[name] == nil {
+		k.existingSecretKeysByNamespaceName[name] = sets.String{}
+	}
+	for key := range item {
+		if key == vault.SecretSyncTargetNamepaceKey || key == vault.SecretSyncTargetNameKey {
+			continue
+		}
+		k.existingSecretKeysByNamespaceName[name].Insert(key)
+		k.existingSecretKeysByVaultSecretName[path] = append(k.existingSecretKeysByVaultSecretName[path], namespacedNameKey{name: name, key: key})
+	}
+}
+
+func (k *kvUpdateTransport) validateKeysDontConflict(ctx context.Context, data map[string]string) (validationErrs []string, err error) {
+	if k.privilegedVaultClient == nil {
+		return nil, nil
+	}
+	k.existingSecretKeysByNamespaceNameLock.Lock()
+	defer k.existingSecretKeysByNamespaceNameLock.Unlock()
+	if k.existingSecretKeysByNamespaceName == nil {
```

Comment:
This whole `if` looks like a good delimitation for an independent function.

Reply:
Do we have an idea of how long this initialization takes?  Can it make the first request time out?  Would it make sense to do it (synchronously or not) at startup?

Reply:
Factored it out and made it run during startup. Not sure how long it takes and depends on the amount of data in Vault. But its definitely nicer to run this outside of user requests.